Substreams-powered subgraphs FAQ (#444)
graphprotocol/docs: website/pages/en/developing/substreams-powered-subgraphs-faq.mdx

---
title: Substreams-powered subgraphs FAQ
---

## What are Substreams?

> Developed by [StreamingFast](https://www.streamingfast.io/), Substreams is an exceptionally powerful processing engine capable of consuming rich streams of blockchain data. Substreams allow you to refine and shape blockchain data for fast and seamless digestion by end-user applications. More specifically, Substreams is a blockchain-agnostic, parallelized, and streaming-first engine, serving as a blockchain data transformation layer. Powered by the [Firehose](https://firehose.streamingfast.io), it ​​enables developers to write Rust modules, build upon community modules, provide extremely high-performance indexing, and [sink](https://substreams.streamingfast.io/developers-guide/sink-targets#substreams-sinks-overview) their data anywhere.

Reviewer: We should be linking to our own "mirror" of the Substreams docs, no?

```suggestion
Developed by [StreamingFast](https://www.streamingfast.io/), Substreams is an exceptionally powerful processing engine capable of consuming rich streams of blockchain data. Substreams allow you to refine and shape blockchain data for fast and seamless digestion by end-user applications. More specifically, Substreams is a blockchain-agnostic, parallelized, and streaming-first engine, serving as a blockchain data transformation layer. Powered by the [Firehose](https://firehose.streamingfast.io), it ​​enables developers to write Rust modules, build upon community modules, provide extremely high-performance indexing, and [sink](/substreams/developers-guide/sink-targets/README/#substreams-sinks-overview) their data anywhere.
```

> Go to the [Substreams Documentation](https://substreams.streamingfast.io/) to learn more about Substreams.

Reviewer: ```suggestion
Go to the [Substreams Documentation](/substreams/README/) to learn more about Substreams.
```

## What are Substreams-powered subgraphs?

> [Substreams-powered subgraphs](/cookbook/substreams-powered-subgraphs/) combine the power of Substreams with the queryability of subgraphs. When publishing a Substreams-powered Subgraph, the data produced by the Substreams transformations, [formatted in a special way](https://github.com/streamingfast/substreams-sink-entity-changes/blob/develop/substreams-entity-change/src/tables.rs), are inserted into the database as if it was produced by the subgraph itself.

Reviewer 2: I am not sure "formatted in a special way" is the best wording, maybe we can leverage the description from the opening of the cookbook? https://thegraph.com/docs/en/cookbook/substreams-powered-subgraphs/

> [Substreams-powered subgraphs](../cookbook/substreams-powered-subgraphs/) combine the power of Substreams with the queryability of subgraphs. When publishing a Substreams-powered Subgraph, the data produced by the Substreams transformations, [formatted in a special way](https://github.com/streamingfast/substreams-sink-entity-changes/blob/develop/substreams-entity-change/src/tables.rs), are inserted into the database as if it was produced by the subgraph itself.

Reviewer: I suggest avoiding relative links, so that they don't break if this page happened to be moved:

```suggestion
[Substreams-powered subgraphs](/cookbook/substreams-powered-subgraphs/) combine the power of Substreams with the queryability of subgraphs. When publishing a Substreams-powered Subgraph, the data produced by the Substreams transformations, [formatted in a special way](https://github.com/streamingfast/substreams-sink-entity-changes/blob/develop/substreams-entity-change/src/tables.rs), are inserted into the database as if it was produced by the subgraph itself.
```

> If you are already familiar with subgraph development, then note that Substreams-powered subgraphs can then be queried, just as if it had been produced by the AssemblyScript transformation layer, with all the Subgraph benefits, like time-travel queries.

Reviewer 2: I don't think we necessarily want to call out time-travel queries as a primary subgraph benefit (I would emphasize the dynamic and flexible GraphQL API)

Author: Good call - fixed that

> ## How are Substreams-powered subgraphs different from subgraphs?

Reviewer 2: I would answer this differently:
Subgraphs are made up of datasources which specify on-chain events, and how those events should be transformed via handlers written in Assemblyscript. These events are processed sequentially, based on the order in which events happen on-chain.

By contrast, substreams-powered subgraphs have a single datasource which references a substreams package, which is processed by the Graph Node. Substreams have access to additional granular on-chain data compared to conventional subgraphs, and can also benefit from massively parallelised processing, which can mean much faster processing times.

Both are used to index and query data from blockchains. They differ in how data transformations are handled. Subgraphs use AssemblyScript to allow you to model your data, whereas Substreams use smaller modules, coded in Rust. Both platforms target a WebAssembly virtual machine.

Another major difference is that Substreams are massively parallelizable, whereas subgraphs need to run linearly across the blockchain history. This parallelization affects the programming paradigm, such that map and store modules are introduced in Substreams to help compute aggregations, whereas Postgres are used in subgraphs.

## What are the benefits of using Substreams-powered subgraphs?

> Substreams-powered subgraphs combine all the benefits of Substreams with the queryability of subgraphs. They bring greater composability and high-performance indexing to The Graph. They also enable new data use cases; for example, once you've built your Substreams-powered Subgraph, you can reuse your [Substreams modules](/substreams/developers-guide/creating-your-manifest/#module-definitions) to output to different [sinks](/substreams/developers-guide/sink-targets/#substreams-sinks-overview) such as PostgreSQL, MongoDB, and Kafka.

Reviewer 2: I think it is worth mentioning the additional data access (e.g. storage changes) as a benefit of substreams

> Substreams-powered subgraphs combine all the benefits of Substreams with the queryability of subgraphs. They bring greater composability and high-performance indexing to The Graph. They also enable new data use cases; for example, once you've built your Substreams-powered Subgraph, you can reuse your [Substreams modules](https://substreams.streamingfast.io/developers-guide/creating-your-manifest#module-definitions) to output to different [sinks](https://substreams.streamingfast.io/developers-guide/sink-targets#substreams-sinks-overview) such as PostgreSQL, MongoDB, and Kafka.

Reviewer: ```suggestion
Substreams-powered subgraphs combine all the benefits of Substreams with the queryability of subgraphs. They bring greater composability and high-performance indexing to The Graph. They also enable new data use cases; for example, once you've built your Substreams-powered Subgraph, you can reuse your [Substreams modules](/substreams/developers-guide/creating-your-manifest#module-definitions) to output to different [sinks](/substreams/developers-guide/sink-targets#substreams-sinks-overview) such as PostgreSQL, MongoDB, and Kafka.
```

## What are the benefits of Substreams?

> There are many benefits to using Substreams, including:

Reviewer 2: Access to additional data which is not available as part of the JSON RPC (this is maybe included in "all the benefits of firehose", but I think it is worth mentioning explicitly)

Go to the [documentation](https://firehose.streamingfast.io/) to learn more about the Firehose.

## What are the benefits of the Firehose?

There are many benefits to using Firehose, including:

- Lowest latency & no polling: In a streaming-first fashion, the Firehose nodes are designed to race to push out the block data first.

- Prevents downtimes: Designed from the ground up for High Availability.

- Never miss a beat: The Firehose stream cursor is designed to handle forks and to continue where you left off in any condition.

- Richest data model:  Best data model that includes the balance changes, the full call tree, internal transactions, logs, storage changes, gas costs, and more.

- Leverages flat files: Blockchain data is extracted into flat files, the cheapest and most optimized computing resource available.

> ## Where can devs access more information about Substreams-powered subgraphs and Substreams?

Reviewer 2: I have a moderately strong preference for developers (not "devs")

Author: Looks more professional - fixed that!

> The [Substreams documentation](/substreams/README/) will teach you how to do Substreams programming.

Reviewer 2: "will teach you how to build Substreams modules"

Author: Good catch! Much better now

> The [Substreams documentation](https://substreams.streamingfast.io) will teach you how to do Substreams programming.

Reviewer: ```suggestion
The [Substreams documentation](/substreams/README/) will teach you how to do Substreams programming.
```

> The [Substreams-powered subgraphs documentation](../cookbook/substreams-powered-subgraphs/) will show you how to package them for deployment on The Graph.

Reviewer: ```suggestion
The [Substreams-powered subgraphs documentation](/cookbook/substreams-powered-subgraphs/) will show you how to package them for deployment on The Graph.
```

## What is the role of Rust modules in Substreams?

Rust modules are the equivalent of the AssemblyScript mappers in subgraphs. They are compiled to WASM in a similar way, but the programming model allows for parallel execution. They define the sort of transformations and aggregations you want to apply to the raw blockchain data.

> See [modules documentation](https://substreams.streamingfast.io/developers-guide/modules) for details.

Reviewer: ```suggestion
See [modules documentation](/substreams/developers-guide/modules/types/) for details.
```

I believe we should have a [`/substreams/developers-guide/modules/README`](https://github.com/streamingfast/substreams/blob/develop/docs/developers-guide/modules/README.md) route but it is currently a 404. @saihaj and @B2o5T, looks like [it is explicitly filtered out](https://github.com/graphprotocol/docs/blob/main/website/scripts/fetch-remote-filepaths.ts#L52)... could it be because the content is the same as the [`/substreams/concepts-and-fundamentals/modules`](https://github.com/streamingfast/substreams/blob/develop/docs/concepts-and-fundamentals/modules.md) page? Is that intentional or an oversight? Maybe we should link to that instead?

Reviewer 3: I think this was intentional cause of similar content. But good idea that we add a redirect

Reviewer: Not similar though, it's exactly the same haha. I meant more, is it intentional that these two (distinct) pages from the Substreams docs have the same content? Maybe one page can be removed.

## What makes Substreams more composable?

> Using subgraphs, you would write mappers in AssemblyScript and deploy those mappers to the network. The subgraph would then index directly to the database, leaving only query-time composability: query two subgraphs, and build your application from those two sources.

Reviewer 2: "Using subgraphs, you would write mappers in AssemblyScript and deploy those mappers to the network" this is slightly confusing phrasing.

I would prefer it if this question was less in comparison to Subgraphs, and more a general FAQ about "What makes Substreams composable?"
Where the answer is that cached modules can be re-used. We haven't currently "solved" how this composability should be discovered and leveraged by users, so I don't think we should over-emphasize it here

Conversely, when using Substreams, the composition happens at the transformation layer. As an example, Alice can build a DEX price module, Bob can use it to build a volume aggregator for some tokens of his interest, and Lisa can combine four individual DEX price modules to create a price oracle. A single Substreams request will package all of these individual's modules, link them together, to offer a much more refined stream of data. That stream can then be used to populate a subgraph, and be queried by consumers.

## How can you build and deploy a Substreams-powered Subgraph?

> After [defining](../cookbook/substreams-powered-subgraphs/) a Substreams-powered Subgraph, you can use the Graph CLI to deploy it in [Subgraph Studio](https://thegraph.com/studio/).

Reviewer: ```suggestion
After [defining](/cookbook/substreams-powered-subgraphs/) a Substreams-powered Subgraph, you can use the Graph CLI to deploy it in [Subgraph Studio](https://thegraph.com/studio/).
```
